index.src.html

```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
+
+  Any <a>reporting observer</a> that is in a <a>registered reporting observer
+  list</a> is considered <dfn>registered</dfn>.
+
+  Each <a>environment settings object</a> has a <dfn>report buffer</dfn>, which
+  is an ordered list of <a>reports</a> that have been generated in
+  that <a>environment settings object</a>. This list is initially empty, and the
+  reports are stored in the same order in which they are generated.
+
+  <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
+
+  <pre class="idl">
+[NoInterfaceObject]
+interface ReportBody {
+};
+
+[NoInterfaceObject]
+interface Report {
+  readonly attribute DOMString type;
+  readonly attribute DOMString url;
+  readonly attribute ReportBody? body;
+};
+
+[Constructor(ReportingObserverCallback callback, optional ReportingObserverOptions options)]
+interface ReportingObserver {
+  void observe();
+  void disconnect();
+};
+
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+  boolean futureOnly = false;
+};
+  </pre>
+
+  Issue: How do we specify which reports are observable?  Network-layer reports
+  like CSP reports will not be.  Maybe there's something about the settings
+  object?
+
+  A <dfn id=dom-report interface>Report</dfn> is the application exposed
+  representation of a <a>report</a>. <dfn attribute for="Report">type</dfn>
+  returns [=report/type=], <dfn attribute for="Report">url</dfn> returns
+  [=report/url=], and <dfn attribute for="Report">body</dfn> returns
+  [=report/body=].
+
+  Each {{ReportingObserver}} object has these associated concepts:
+    - A <dfn for=ReportingObserver>callback</dfn> function set on creation.
+    - A {{ReportingObserverOptions}} dictionary called
+      <dfn for=ReportingObserver>options</dfn>.
+    - A list of {{Report}} objects called the <dfn for=ReportingObserver>report
+      list</dfn>, which is initially empty.
+
+  The <dfn constructor for=ReportingObserver><code>
+  ReportingObserver(|callback|, |options|)</code></dfn> constructor, when invoked, must run
+  these steps:
+
+  1. Create a new {{ReportingObserver}} object |observer|.
+
+  2. Set |observer|'s <a>callback</a> to |callback|.
+
+  3. Set |observer|'s <a>options</a> to |options|.
+
+  4. Return |observer|.
+
+  The <dfn method for=ReportingObserver><code>observe()</code></dfn>
+  method, when invoked, must run these steps:
+
+  1. Let |settings| be the <a>environment settings object</a> associated with
+     the <a>context object</a>.
+
+  2. Append the <a>context object</a> to the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  3. If the <a>context object</a>'s {{futureOnly}} <a>option</a> is true, return.
+
+  4. Set <a>context object</a>'s {{futureOnly}} <a>option</a> to true.
+
+  5. For each |report| in the <a>report buffer</a> associated with |settings|,
```

Comment:
This implies that we must queue all reports indefinitely for the lifetime of the page. Is that the behavior we want, and/or reason for concern? It's not crazy to imagine hundreds of reports sitting around indefinitely..

Reply:
This is so that observers can be made to observe reports that were generated before they could be registered, and this is the behavior we want. We don't expect that the number of reports will be that large, but there is also the buffer limit of 100 in place, so there can't be hundreds sitting around.

Reply:
There's a section further down about garbage collection, which gives the user agent the ability to throw away reports before they're delivered if they've been sitting around too long.  I suggest updating that text to say that if any reports are ever GCed from the reporting cache, they must also be removed from any observer's report list that they appear in.